File: Python/112. Path Sum.py
'''
1.递归
2.回溯
'''


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# class Solution:
#     def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
#         if not root:
#             return False
#         sum_ = 0
#         def traversal(root, targetSum):
#             nonlocal sum_
#             sum_ += root.val
#             if not root.left and not root.right and sum_ == targetSum:
#                 return True
#             if root.left and traversal(root.left, targetSum):
#                 return True
#             if root.right and traversal(root.right, targetSum):
#                 return True
#             sum_ -= root.val
#             return False
#         return traversal(root, targetSum)

class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False
        st = [(root, root.val)]
        while st:
            cur = st.pop(-1)
            if not cur[0].left and not cur[0].right and cur[1] == targetSum:
                return True
            if cur[0].right:
                st.append((cur[0].right, cur[0].right.val + cur[1]))
            if cur[0].left:
                st.append((cur[0].left, cur[0].left.val + cur[1]))
        return False

File: C++/112. Path Sum.cpp
/*
1.µÝ¹é
2.µü´ú
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 // class Solution {
 // public:
 //     int sum = 0;
 //     bool hasPathSum(TreeNode* root, int targetSum) {
 //         if (root == NULL){
 //             return false;
 //         }
 //         sum += root->val;
 //         if (!root->left and !root->right){
 //             if (sum == targetSum) return true;
 //         }
 //         if (root->left){
 //             if (hasPathSum(root->left, targetSum)) return true;
 //         }
 //         if (root->right){
 //             if (hasPathSum(root->right, targetSum)) return true;
 //         }
 //         sum -= root->val;
 //         return false;
 //     }
 // };

class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        if (root == NULL) return false;
        stack<pair<TreeNode*, int>>st;
        st.push(pair<TreeNode*, int>(root, root->val));
        while (!st.empty()) {
            pair<TreeNode*, int> cur = st.top(); st.pop();
            if (!cur.first->left && !cur.first->right && cur.second == targetSum) return true;
            if (cur.first->right) {
                st.push(pair<TreeNode*, int>(cur.first->right, cur.second + cur.first->right->val));
            }
            if (cur.first->left) {
                st.push(pair<TreeNode*, int>(cur.first->left, cur.second + cur.first->left->val));
            }
        }
        return false;
    }
};